synapse/lib/filepath.py

```diff
@@ -0,0 +1,371 @@
+        self.zippath = path
+
+        super(FilePathZip, self).__init__(zipfd, '/', remainder, parent=parent)
+
+    def _type(self):
+        return 'zip.file'
+
+    def _open(self, mode='r'):
+        return open(self.zippath, mode=mode)
+
+def zip_enumerate(zipfd):
+    '''
+    Return the files and dirs in a zip file
+    '''
+    files = set()
+    dirs = set()
+
+    for path in zipfd.namelist():
+        
+        for spath in _get_subpaths(path):
+            try:
+                info = zipfd.getinfo(spath.strip('/'))
+                files.add(spath)
+            except KeyError as e:
+                dirs.add(spath)
+
+    return dirs, files
+
+def tar_enumerate(tarfd):
+    '''
+    Return the files and dirs in a tar archive
+    '''
+    files = set()
+    dirs = set()
+
+    for path in tarfd.getnames():
+        
+        for spath in _get_subpaths(path):
+            info = tarfd.getmember(spath.strip('/'))
+            if info.isdir():
+                dirs.add(spath)
+            if info.isfile():
+                files.add(spath)
+                
+    return dirs, files
+
+def _get_subpaths(path):
+    '''
+    Returns a list of subpaths in a path, one for each level in the path
+    '''
+    path_parts = path.strip('/').split('/')
+
+    paths = []
+    for i in range(len(path_parts)):
+
+        ipath = os.path.join('/', *path_parts[:i+1])
+        paths.append(ipath)
+
+    return paths
+
+def _enum_path(path):
+    fpobj = _parse_path(path)
+
+    if fpobj:
+        return fpobj._type()
+    return None
+
+def _parse_path(path):
+    '''
+    Internal function to parse the incoming path
+    '''
+
+    if not path:
+        return None
+
+    path = genpath(path)
+
+    path_parts = path.strip('/').split('/')
+
+    fpobj = None
+    next_path = os.path.join('/', path_parts[0])
+    remainder = '/'.join(path_parts[1:])
+
+    try:
+        cls = _get_path_class(next_path)
+        fpobj = cls(next_path, remainder)
+
+        while fpobj.getChild():
+            fpobj = fpobj.getChild()
+    except s_exc.NoSuchPath as e:
+        return None
+
+    return fpobj
+
+def _open(path, mode='r'):
+    '''
+    Returns a read-only file-like object even is the path terminates inside a container file.
+    If the path is a regular os accessible path mode is passed through.  If the path terminates 
+    in a container file mode is ignored.
+
+    If the path does not exist a NoSuchPath exception is raised.
+
+    ex.
+    _open('/foo/bar/baz.egg/path/inside/zip/to/file')
+    '''
+    fpobj = _parse_path(path)
+    if not fpobj:
+        raise s_exc.NoSuchPath(path=path)
+    if not fpobj._type().endswith('.file'):
+        raise s_exc.NoSuchPath(path=path)
+
+    fd = fpobj._open(mode=mode)
+    return fd
+
+def isfile(path):
+    '''
+    Determines if the path is a file, even if the path terminates inside a container file.
+    Returns a boolean.
+    '''
+    ptype = _enum_path(path)
+    if ptype and ptype.endswith('.file'):
```

Comment:
No, it can't be a file and a directory.  It's one or the other.  I'll add an isfile to the base object.  If it returns False the assumption is that it's a directory.